pkg/internal/util/cli/logger.go

```diff
@@ -20,58 +20,105 @@ import (
 	"bytes"
 	"fmt"
 	"io"
+	"runtime"
+	"strings"
 	"sync"
 
 	"sigs.k8s.io/kind/pkg/log"
 )
 
 // Logger is the kind cli's log.Logger implementation
 type Logger struct {
-	Verbosity log.Level
-	io.Writer
-	writeMu sync.Mutex
+	writer     io.Writer
+	writerMu   sync.Mutex
+	verbosity  log.Level
+	bufferPool *bufferPool
 }
 
 var _ log.Logger = &Logger{}
 
 // NewLogger returns a new Logger with the given verbosity
 func NewLogger(writer io.Writer, verbosity log.Level) *Logger {
 	return &Logger{
-		Verbosity: verbosity,
-		Writer:    writer,
+		verbosity:  verbosity,
+		writer:     writer,
+		bufferPool: newBufferPool(),
 	}
 }
 
-func (l *Logger) Write(p []byte) (n int, err error) {
-	// TODO: line oriented instead?
-	// For now we make a single per-message write call from the rest of the logger
-	// intentionally to effectively do this one level up
-	l.writeMu.Lock()
-	defer l.writeMu.Unlock()
-	return l.Writer.Write(p)
+// synchronized write to the inner writer
+func (l *Logger) write(p []byte) (n int, err error) {
+	l.writerMu.Lock()
+	defer l.writerMu.Unlock()
+	return l.writer.Write(p)
 }
 
-// TODO: prefix log lines with metadata (log level? timestamp?)
-
-func (l *Logger) print(message string) {
-	buf := bytes.NewBufferString(message)
+// writeBuffer writes buf with write, ensuring there is a trailing newline
+func (l *Logger) writeBuffer(buf *bytes.Buffer) {
+	// ensure trailing newline
 	if buf.Bytes()[buf.Len()-1] != '\n' {
 		buf.WriteByte('\n')
 	}
 	// TODO: should we handle this somehow??
 	// Who logs for the logger? 🤔
-	_, _ = l.Write(buf.Bytes())
+	_, _ = l.write(buf.Bytes())
+}
+
+// print writes a simple string to the log writer
+func (l *Logger) print(message string) {
+	buf := bytes.NewBufferString(message)
```

Comment:
This one does not use a buffer pool currently because this is basically just a cast:
https://github.com/golang/go/blob/3a1b4e75f8b6c1b57db73bccf7ca871bf1a97ca9/src/bytes/buffer.go#L460

However, adding the newline will still frequently cause an allocation. We could write a more clever version of this in a future version, but I'm doubtful that it matters at the moment..

 The most important performance things about this logger while we target CLI logging is just that `V(99999999)` is cheap when `-v=0` and that we synchronize output (no interleaved lines). Not wasting allocations is just nice when it's relatively trivial :-)

Reply:
Otoh, `fmt.Sprintf` internally does some pooling, as does `k8s.io/klog`, so it made sense to slap in the trivial pool for the `*f` method family ...

Reply:
we're now mimicing klog and dropping buffers > 256 bytes instead of retaining them